### PR TITLE
Exact conversions from `Long` to `Int`

### PR DESCRIPTION
**Pull request by CommanderTvis**
resolves #374

### CODE REVIEW

File: kmath-dimensions/src/commonMain/kotlin/space/kscience/kmath/dimensions/Wrappers.kt [hunk 1]
```diff
@@ -23,11 +24,11 @@ public interface DMatrix<T, R : Dimension, C : Dimension> : Structure2D<T> {
          * Coerces a regular matrix to a matrix with type-safe dimensions and throws a error if coercion failed
          */
         public inline fun <T, reified R : Dimension, reified C : Dimension> coerce(structure: Structure2D<T>): DMatrix<T, R, C> {
-            require(structure.rowNum == Dimension.dim<R>().toInt()) {
+            require(structure.rowNum == Dimension.dim<R>().toIntExact()) {
```
Comment (altavir) [inline kmath-dimensions/src/commonMain/kotlin/space/kscience/kmath/dimensions/Wrappers.kt:27]: Maybe it is better to use UInt for rowNumbers?

File: kmath-nd4j/src/main/kotlin/space/kscience/kmath/nd4j/arrays.kt [hunk 2]
```diff
@@ -5,4 +5,6 @@
 
 package space.kscience.kmath.nd4j
 
-internal fun LongArray.toIntArray(): IntArray = IntArray(size) { this[it].toInt() }
+import space.kscience.kmath.misc.toIntExact
+
+internal fun LongArray.toIntArray(): IntArray = IntArray(size) { this[it].toIntExact() }
```
Comment (altavir) [inline kmath-nd4j/src/main/kotlin/space/kscience/kmath/nd4j/arrays.kt:10]: Is it used somewhere? Does not seem to be safe.

File: kmath-stat/src/commonMain/kotlin/space/kscience/kmath/samplers/PoissonSampler.kt [hunk 3]
```diff
@@ -52,11 +53,10 @@ public class SmallMeanPoissonSampler(public val mean: Double) : Sampler<Int> {
 
     private val p0: Double = exp(-mean)
 
-    private val limit: Int = if (p0 > 0) {
+    private val limit: Int = (if (p0 > 0)
```
Comment (altavir) [inline kmath-stat/src/commonMain/kotlin/space/kscience/kmath/samplers/PoissonSampler.kt:56]: Whi bracket here? Does not look nice.

File: kmath-dimensions/src/commonMain/kotlin/space/kscience/kmath/dimensions/Wrappers.kt (hunk 1 above)
Comment (CommanderTvis) [reply kmath-dimensions/src/commonMain/kotlin/space/kscience/kmath/dimensions/Wrappers.kt:27]: `Structure2D.rowNum: UInt`?

File: kmath-nd4j/src/main/kotlin/space/kscience/kmath/nd4j/arrays.kt (hunk 2 above)
Comment (CommanderTvis) [reply kmath-nd4j/src/main/kotlin/space/kscience/kmath/nd4j/arrays.kt:10]: It is used for converting ND4J `long[]` indices to KMath's `int[]`.

File: kmath-stat/src/commonMain/kotlin/space/kscience/kmath/samplers/PoissonSampler.kt (hunk 3 above)
Comment (CommanderTvis) [reply kmath-stat/src/commonMain/kotlin/space/kscience/kmath/samplers/PoissonSampler.kt:56]: Reverted

File: kmath-dimensions/src/commonMain/kotlin/space/kscience/kmath/dimensions/Wrappers.kt (hunk 1 above)
Comment (altavir) [reply kmath-dimensions/src/commonMain/kotlin/space/kscience/kmath/dimensions/Wrappers.kt:27]: yes. It does not make sense to use Ints now that UInt is stable

Comment (CommanderTvis) [reply kmath-dimensions/src/commonMain/kotlin/space/kscience/kmath/dimensions/Wrappers.kt:27]: Should I try to do it in this branch?

Comment (altavir) [reply kmath-dimensions/src/commonMain/kotlin/space/kscience/kmath/dimensions/Wrappers.kt:27]: Yes, I think so

**Comment by altavir** [comment]
Need to resolve conflict

**Comment by CommanderTvis** [comment]
> Need to resolve conflict

Done.

**Review by altavir — APPROVED** [review]

**Pull request merged**